proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
 fun addInts(a: Int, b: Int): Int = with(intext.IntMonoid) { add(a, b) }
 ```
 
-## Compile resolution rules
+## Type Class Instance Rules
 
-When the compiler finds a call site invoking a function that has type class instances constrains declared with `with` as in the example below:
+Classical interfaces only allow the implementation of interfaces to occur when a type is defined. Type classes typically relax this rule and allow implementations outside of the type definition. When relaxinng this rule it is important to preserve the coherency we take for granted with classical interfaces.
 
-Declaration: 
+For those reasons type class instances must be declared in one of two places:
+
+1. In the same file as the type class definition (interface-side implementation)
+2. In the same file as the type being implemented (type-side implementation)
+
+All other instances are orphan instances and are not allowed. See [Appendix A](#Appendix-A) for a modification to this proposal that allows for orphan instances.
```

Comment:
Overriding default behaviors specially in effects libraries like `arrow-effect` is important. I think internal overrides should be part of this proposal and not separate. In the OOP side of things in Kotlin they are gonna want to override instances for many purposes including tests, swapping runtime implementations, etc.

Reply:
On the OOP side of things you override behavior by defining a new class, you are not allowed to change the behavior of an already defined class. It's the same thing with type classes: you can define a new type that has different type class behavior.

If you program against the type class the same way you program against a classical instance, you can swap out whichever concrete types you want.

Adding the ability to dynamically modify interface implementation might be interesting, but I would recommend starting with something conservative. Adding features is always easier than removing them.

Reply:
I will add that type classes that want to be "overrideable" can be designed that way. Imagine a type class for JSON serialization. We want to provide multiple instances for different locations. We can do that by use of a "phantom" type class parameter to control selection. This also has the benefit of forcing us to select the implementation we want and having the compiler guarantee consistency.

```kotlin
object Production
object Testing

extension interface FromJson<T, Marker> {
   fun fromJson(j: JsonNode): T
}
object : FromJson<Int, Production> { ... }
object : FromJson<Int, Testing> { ... }

FromJson<Int, Production>.fromJson(...)
```

Reply:
I like that, thanks for the example. I Imagine cases like `MonadError<F, E>` are addressed in a similar way.